### PR TITLE
drivers: serial: nrf uarte: avoid dropping RX chars/overruns

### PR DESCRIPTION
**Pull request by marc-cpdesign**
In some cases (eg at high baud rate, no HW flow control, and when BLE
radio/ints running) data could be lost between when enough characters
have been RX'd to fill the DMA buffer and when the ENDRX event was fired,
where the the STARTRX task is invoked to start filling the next buffer
(which is set up earlier, but I think will not be filled until STARTRX).
To fix this, the SHORT is enabled between ENDRX and STARTRX whenever the
'next' buffer is available, so that STARTRX is invoked automatically and
subsequent chars go into the next buffer via EasyDMA.
To make this work properly, uarte_nrfx_isr_async() now handles the ENDRX
event _before_ the STARTRX event.

There was also an issue in rx_timeout() where the received character
count (rx_total_byte_count) could be incremented greater than the actual
buffer size. This arises from rx_total_byte_count value coming from the
counting the RXDRDY events (either by PPI/timer counter or counting the
RXDRDY ints themselves) and so if chars are received in the rx_timeout()
(or before ENDRX is handled) the rx_timeout() could increment rx_offset
past the length of the buffer. This could result the remaining 'len' being
calculated incorrectly (an underflow due to unsigned - signed , where
signed > unsigned).
To fix this, we now store the lengths of the buffers and don't invoke the
UART_RX_RDY callback when the buffers are full; its handled by ENDRX.
(Also note that the buffer size should be available via the RXD.MAXCNT
register on the nrf, but this register is not exposed through the nrfx HAL
and is also double buffered, so it seemed clearer to just track the buffer
lengths explicitly here in the driver).

Signed-off-by: Marc Reilly <marc@cpdesign.com.au>

### CODE REVIEW

**Comment by cybertale** [comment]
Hi, you can check the CODEOWNER file and see whom files you edited belong to, then add them as reviwers manually or @their names.

**Comment by marc-cpdesign** [comment]
Hi Song,

On Thursday, 9 April 2020 12:06:29 PM AEST Song Qiang wrote:
> Hi, you can check the CODEOWNER file and see whom files you edited belong
> to, then add them as reviwers manually or @their names.

Thanks, but in this case there is no matching entry for /drivers/serial/
uart_nrfx_uarte.c

Cheers
Marc

**Comment by cybertale** [comment]
Actually the owner of the driver folder , the uart driver folder and maintainer of nrf socs all can review this, especially the latter two.

**Comment by rerickson1** [comment]
When this is ready, I would ask that this is back-ported to 1.14.x, thanks!

**Comment by carlescufi** [comment]
Both @anangl and @Mierunski need to look at this before we merge.
Regarding the backport, I will add a label, but someone needs to send the PR

**Comment by rerickson1** [comment]
#24245 covers backport to 1.14.x

File: drivers/serial/uart_nrfx_uarte.c
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte.c:759]: This comment may be misleading (at least I was a bit confused by it). `rx_offset` can be bigger than `this_amount` because it can also include some data from the next buffer that is already being filled up by the transfer started via the ENDRX_STARTRX shortcut.

**Review by anangl — APPROVED** [review]
Good job, @marc-cpdesign. Thank you for providing this improvement. And my apologies that you had to wait so long for a review. I'd suggest to only rephrase one comment, but maybe it's only me that couldn't understand it well, so I won't insist on that.

File: drivers/serial/uart_nrfx_uarte.c
Comment (Mierunski) [inline drivers/serial/uart_nrfx_uarte.c:539]: Better to use NULL for pointer.

File: drivers/serial/uart_nrfx_uarte.c
Comment (Mierunski) [inline drivers/serial/uart_nrfx_uarte.c:754]: `this_amount` seems a bit ambiguous, maybe `rx_amount` ?

**Review by Mierunski — CHANGES_REQUESTED** [review]
Looks good overall, just small suggestions

File: drivers/serial/uart_nrfx_uarte.c
Comment (marc-cpdesign) [reply drivers/serial/uart_nrfx_uarte.c]: fixed

File: drivers/serial/uart_nrfx_uarte.c
Comment (marc-cpdesign) [reply drivers/serial/uart_nrfx_uarte.c]: I changed to rx_amount

File: drivers/serial/uart_nrfx_uarte.c
Comment (marc-cpdesign) [reply drivers/serial/uart_nrfx_uarte.c]: updated the comment.

**Review by anangl — APPROVED** [review]

**Review by Mierunski — APPROVED** [review]

**Pull request merged**
